Google.Solutions.Compute/Iap/SshRelayListener.cs

```diff
@@ -0,0 +1,184 @@
+        public event EventHandler<ClientEventArgs> ClientConnected;
+        public event EventHandler<ClientEventArgs> ClientDisconnected;
+        public event EventHandler<ConnectionFailedEventArgs> ConnectionFailed;
+
+        public int ClientsAccepted { get; private set; } = 0;
+
+        // For testing only.
+        internal int ClientAcceptLimit { get; set; } = 0;
+
+        public class ClientEventArgs : EventArgs
+        {
+            public string Client { get; }
+
+            public ClientEventArgs(string client)
+            {
+                this.Client = client;
+            }
+        }
+
+        public class ConnectionFailedEventArgs : EventArgs
+        {
+            public Exception Exception { get; }
+
+            public ConnectionFailedEventArgs(Exception exception)
+            {
+                this.Exception = exception;
+            }
+        }
+
+        //---------------------------------------------------------------------
+        // Ctor
+        //---------------------------------------------------------------------
+
+        private SshRelayListener(ISshRelayEndpoint server, int localPort)
+        {
+            this.server = server;
+            this.LocalPort = localPort;
+
+            this.listener = new TcpListener(new IPEndPoint(IPAddress.Loopback, localPort));
+        }
+
+        //---------------------------------------------------------------------
+        // Events
+        //---------------------------------------------------------------------
+
+        private void OnClientConnected(string client)
+        {
+            if (this.ClientConnected != null)
+            {
+                this.ClientConnected(this, new ClientEventArgs(client));
+            }
+        }
+        private void OnClientDisconnected(string client)
+        {
+            if (this.ClientDisconnected != null)
+            {
+                this.ClientDisconnected(this, new ClientEventArgs(client));
+            }
+        }
+        private void OnConnectionFailed(Exception e)
+        {
+            if (this.ConnectionFailed != null)
+            {
+                this.ConnectionFailed(this, new ConnectionFailedEventArgs(e));
+            }
+        }
+
+        //---------------------------------------------------------------------
+        // Publics
+        //---------------------------------------------------------------------
+
+        /// <summary>
+        ///  Create a listener using a dynamically selected, unused local port.
+        /// </summary>
+        public static SshRelayListener CreateLocalListener(ISshRelayEndpoint server)
+        {
+            return CreateLocalListener(server, PortFinder.FindFreeLocalPort());
+        }
+
+        /// <summary>
+        ///  Create a listener using a defined local port.
+        /// </summary>
+        public static SshRelayListener CreateLocalListener(ISshRelayEndpoint server, int port)
+        {
+            if (port < 0 || port > ushort.MaxValue)
```

Comment:
its prudent to check the port ranges here ...but why that range? is it if custoemrs define a port on their own (vs above where w'ere iterating over within .FindFreeLocalPort()?


(i think if you ask binding to port '0' its dynamically allocated,)

Reply:
This is simply because of the int vs ushort mismatch. Ports are ushort, but the API uses int.